libs/ui/src/hooks/useHardware.ts

```diff
@@ -0,0 +1,175 @@
+  const [
+    hasAccessibleControllerAttached,
+    setHasAccessibleControllerAttached,
+  ] = useState(false);
+  const [hasPlustekScannerAttached, setHasPlustekScannerAttached] = useState(
+    false
+  );
+  const [hasFujitsuScannerAttached, setHasFujitsuScannerAttached] = useState(
+    false
+  );
+  const [hasPrinterAttached, setHasPrinterAttached] = useState(false);
+  const [allDevices, setAllDevices] = useState<KioskBrowser.Device[]>([]);
+  const [allPrinters, setAllPrinters] = useState<KioskBrowser.PrinterInfo[]>(
+    []
+  );
+  const previousDevices = usePrevious(allDevices);
+  const previousPrinters = usePrevious(allPrinters);
+
+  useEffect(() => {
+    const hardwareStatusSubscription = hardware.devices
+      .pipe(map((devices) => Array.from(devices)))
+      .subscribe(async (devices) => {
+        setHasCardReaderAttached(devices.some(isCardReader));
+        setHasAccessibleControllerAttached(
+          devices.some(isAccessibleController)
+        );
+        setHasPlustekScannerAttached(devices.some(isPlustekVTM300Scanner));
+        setHasFujitsuScannerAttached(devices.some(isFujitsuScanner));
+        setAllDevices(devices);
+      });
+
+    const printerStatusSubscription = hardware.printers
+      .pipe(map((printers) => Array.from(printers)))
+      .subscribe(async (printers) => {
+        const newHasPrinterAttached = printers.some(
+          ({ connected }) => connected
+        );
+        setHasPrinterAttached(newHasPrinterAttached);
+        setAllPrinters(printers);
+      });
+    return () => {
+      hardwareStatusSubscription.unsubscribe();
+      printerStatusSubscription.unsubscribe();
+    };
+  }, [hardware]);
+
+  // Handle logging of printers
+  useEffect(() => {
+    for (const newPrinter of allPrinters) {
+      const previousCopyOfPrinter =
+        previousPrinters &&
+        previousPrinters.find((printer) => printer.name === newPrinter.name);
+      if (!previousCopyOfPrinter) {
+        void logger.log(LogEventId.PrinterAttached, 'system', {
+          message: `New printer attached: ${newPrinter.name} with connection status: ${newPrinter.connected}`,
+          printer: newPrinter.name,
+          connected: newPrinter.connected.toString(),
+        });
+      } else if (previousCopyOfPrinter.connected !== newPrinter.connected) {
+        void logger.log(LogEventId.PrinterConnectionUpdate, 'system', {
+          message: `Printer ${newPrinter.name} has been ${
+            newPrinter.connected ? 'connected' : 'disconnected'
+          }.`,
+          printer: newPrinter.name,
+          connected: newPrinter.connected.toString(),
+        });
+      }
+    }
+    for (const oldPrinter of previousPrinters || []) {
+      const newCopyOfPrinter = allPrinters.find(
+        (printer) => printer.name === oldPrinter.name
+      );
+      if (!newCopyOfPrinter) {
+        void logger.log(LogEventId.PrinterUnattached, 'system', {
```

Comment:
@eventualbuddha Not sure if you are more familiar with how kiosk-browser works for this but I don't fully understand when/if this would happen. In my testing I basically always had two printers (one named for my actual HP printer and the other "VxPrinter" in the array coming from window.kiosk.printers.subscribe)and as I connect/disconnect from my VM the connection status (on both of these) will flip but they never actually get unattached and are present even when the printer is not connected.

Reply:
I assume this would happen if the printer is unconfigured. Like, you go into the system settings and remove the printer from the list.